mlflow/openai/api_request_parallel_processor.py

```diff
@@ -71,283 +59,58 @@ def increment_num_rate_limit_errors(self):
-            if isinstance(prompt, str):  # single prompt
-                prompt_tokens = len(encoding.encode(prompt))
-                return prompt_tokens + completion_tokens
-            elif isinstance(prompt, list):  # multiple prompts
-                prompt_tokens = sum([len(encoding.encode(p)) for p in prompt])
-                return prompt_tokens + completion_tokens * len(prompt)
-            else:
-                raise TypeError(
-                    "Expecting either string or list of strings for 'prompt' field in completion "
-                    "request"
-                )
-    # if embeddings request, tokens = input tokens
-    elif "embeddings" in request_url:
-        inp = request_json["input"]
-        if isinstance(inp, str):  # single input
-            return len(encoding.encode(inp))
-        elif isinstance(inp, list):  # multiple inputs
-            return sum([len(encoding.encode(i)) for i in inp])
-        else:
-            raise TypeError(
-                'Expecting either string or list of strings for "inputs" field in embedding request'
-            )
-    # more logic needed to support other API calls (e.g., edits, inserts, DALL-E)
-    else:
-        raise NotImplementedError(f'Support for "{request_url}" not implemented in this script')
+    import openai
+
+    status_tracker.start_task()
+    try:
+        result = task()
+        _logger.debug(f"Request #{index} succeeded")
+        status_tracker.complete_task(success=True)
+        results.append((index, result))
+    except openai.RateLimitError as e:
+        status_tracker.complete_task(success=False)
+        _logger.debug(f"Request #{index} failed with {e}")
+        status_tracker.increment_num_rate_limit_errors()
+        status_tracker.error = mlflow.MlflowException(
+            f"Request #{index} failed with rate limit: {e}."
+        )
+    except Exception as e:
+        status_tracker.complete_task(success=False)
+        _logger.debug(f"Request #{index} failed with {e}")
+        status_tracker.error = mlflow.MlflowException(f"Request #{index} failed with {e.__cause__}")
 
 
 def process_api_requests(
-    requests: list[dict[str, any]],
-    request_url: str,
-    api_token: _OAITokenHolder,
-    # Reference: https://platform.openai.com/docs/guides/rate-limits/overview
-    max_requests_per_minute: float = 3_500,
-    max_tokens_per_minute: float = 90_000,
-    token_encoding_name: str = "cl100k_base",
-    max_attempts: int = 5,
+    request_tasks: list[Callable[[], Any]],
     max_workers: int = 10,
-    throw_original_error=False,
-    extra_headers: Optional[dict[str, str]] = None,
 ):
-    """Processes API requests in parallel, throttling to stay under rate limits."""
-    # constants
-    seconds_to_pause_after_rate_limit_error = 15
-
+    """Processes API requests in parallel"""
     # initialize trackers
-    retry_queue = queue.Queue()
     status_tracker = StatusTracker()  # single instance to track a collection of variables
```

Comment:
Do we still need `StatusTracker`? If not, can we remove `StatusTracker` in a follow-up?